### PR TITLE
Masked gemv

### PR DESCRIPTION
**Pull request by jagrit06**
## Proposed changes

Comes with general improvements to gemv kernel and specialized masked gemv kernel 

Benchmarks of vanilla gemv on M3 Max Laptop:
Screenshot 2024-06-14 at 8 52 18 AM

Screenshot 2024-06-14 at 8 52 39 AM

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: python/tests/test_blas.py [hunk 1]
```diff
@@ -794,10 +796,14 @@ def f_test(a_, b_):
             out_ref, dout_ref = mx.vjp(f_ref, [a, b], [cotan])
             out_test, dout_test = mx.vjp(f_test, [a, b], [cotan])
 
-            mx.eval((out_ref, dout_ref, out_test, dout_test))
+            # mx.eval((out_ref, dout_ref, out_test, dout_test))
```
Comment (awni) [inline python/tests/test_blas.py:799]: Should we just delete that line?

File: python/tests/test_blas.py [hunk 2]
```diff
@@ -896,28 +902,58 @@ def test_shape(
...
             (256, 256, 128, 64),
+            (1, 128, 128, 32),
+            (256, 1, 128, 64),
         )
 
         for M, N, K, block_size in shapes:
             test_shape(M, N, K, block_size)
 
-        # Test broadcasting
+        # # Test broadcasting
```
Comment (awni) [inline python/tests/test_blas.py:912]: Nit: double comment here

**Review by awni — COMMENTED** [review]
Wow! The results here are incredible!!

**Comment by jagrit06** [comment]
Small benchmarks on masked gemv 

```
  B,     M,     N,     K,   dtype,  t, block_size,  time, time_masked, diff% = 100. * (1.0 - time_mlx_masked / time_mlx)
  1,  1024,     1,  4096, float32, nt,         64, 0.172,       0.129, +25.23%
  1,     1,  1024,  4096, float32, nt,         64, 0.192,       0.104, +45.91%
  1,     1,  4096,  1024, float32, nt,         64, 0.162,       0.107, +34.20%
  1,     1,  4096,  4096, float32, nt,         64, 0.644,       0.141, +78.18%
```

**Review by awni — APPROVED** [review]
🚀 🚀

**Pull request merged**